packages/snaps-controllers/src/cronjob/CronjobController.ts

```diff
@@ -91,529 +83,420 @@ export const DAILY_TIMEOUT = inMilliseconds(24, Duration.Hour);
+      `${controllerName}:get`,
+      (...args) => this.get(...args),
     );
 
-    this.dailyCheckIn();
-
-    this.#rescheduleBackgroundEvents(Object.values(this.state.events));
+    this.#clear();
+    this.#reschedule();
+    this.#start();
   }
 
   /**
-   * Retrieve all cronjob specifications for all runnable snaps.
+   * Schedule an event.
    *
-   * @returns Array of Cronjob specifications.
+   * @param event - The event to schedule.
+   * @returns The ID of the scheduled event.
    */
-  #getAllJobs(): Cronjob[] {
-    const snaps = this.messagingSystem.call('SnapController:getAll');
-    const filteredSnaps = getRunnableSnaps(snaps);
+  schedule(event: SchedulableBackgroundEvent) {
+    const id = event.id ?? nanoid();
+    const internalEvent: InternalBackgroundEvent = {
+      ...event,
+      id,
+      date: getExecutionDate(event.schedule),
+      scheduledAt: getCurrentDate(),
+    };
+
+    this.update((state) => {
+      state.events[internalEvent.id] = castDraft(internalEvent);
+    });
 
-    const jobs = filteredSnaps.map((snap) => this.#getSnapJobs(snap.id));
-    return jobs.flat().filter((job) => job !== undefined) as Cronjob[];
+    this.#schedule(internalEvent);
+    return id;
   }
 
   /**
-   * Retrieve all Cronjob specifications for a Snap.
+   * Cancel an event.
    *
-   * @param snapId - ID of a Snap.
-   * @returns Array of Cronjob specifications.
+   * @param origin - The origin making the cancel call.
+   * @param id - The id of the event to cancel.
+   * @throws If the event does not exist.
    */
-  #getSnapJobs(snapId: SnapId): Cronjob[] | undefined {
-    const permissions = this.messagingSystem.call(
-      'PermissionController:getPermissions',
-      snapId,
+  cancel(origin: string, id: string) {
+    assert(
+      this.state.events[id],
+      `A background event with the id of "${id}" does not exist.`,
     );
 
-    const permission = permissions?.[SnapEndowments.Cronjob];
-    const definitions = getCronjobCaveatJobs(permission);
+    assert(
+      this.state.events[id].snapId === origin,
+      'Only the origin that scheduled this event can cancel it.',
+    );
 
-    return definitions?.map((definition, idx) => {
-      return { ...definition, id: `${snapId}-${idx}`, snapId };
-    });
+    this.#cancel(id);
   }
 
   /**
-   * Register cron jobs for a given snap by getting specification from a permission caveats.
-   * Once registered, each job will be scheduled.
+   * Get a list of a Snap's background events.
    *
-   * @param snapId - ID of a snap.
+   * @param snapId - The id of the Snap to fetch background events for.
+   * @returns An array of background events.
    */
-  register(snapId: SnapId) {
-    const jobs = this.#getSnapJobs(snapId);
-    jobs?.forEach((job) => this.#schedule(job));
+  get(snapId: SnapId): InternalBackgroundEvent[] {
+    return Object.values(this.state.events).filter(
+      (snapEvent) => snapEvent.snapId === snapId, // && !snapEvent.recurring,
```

Comment:
Temporarily commented out for testing purposes, but need to revert this before merging, at least assuming we don't want to include cronjobs in `snap_getBackgroundEvents`.

```suggestion
      (snapEvent) => snapEvent.snapId === snapId && !snapEvent.recurring,
```